### DISCUSSION TITLE
Why a segmentation fault could pop up only above a number of requested MPI processes?

### DISCUSSION THREAD

**Discussion by garciapintado** [idaholab/moose#22813] [Q&A General] [answered]
Hi,
I have an error showing up only when I use over a number of processes. So that in my machine, for “mpiexec -np 15 …” it works while for higher number of mpi processes (>= 16) an error shows up:
```pre
—
…
Time Step 1, time = 1000, dt = 1000

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 1357638 RUNNING AT clg1
=   EXIT CODE: 9
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Segmentation fault (signal 11)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
—-
```
I haven’t used a parallel debugger before, so before trying I am wondering if you have a suggestion about why this could be. I do not use any other object that is not in the default MOOSE distribution [snapshot-20-10-27-26481-g16505c700e], and these are objects within the Porous Flow and Heat Conduction modules.

Also the error is specific to my input files [an imported 3D mesh, which is also used to initiate the nonlinear variables]. It does not reproduce for very similar problems [even with a problem twice as big] where  the mesh and initialisation is done within MOOSE, and I can use ‘mpi -n 24 …’  (the number of physical cores in my machine) without any issue.

So, I’m puzzled about why an error could only show up when the level of parallelisation goes over a threshold, and wondering if have faced a simular situation before.

A note is that the problem is a 3D domain, with several blocks, which result from extrusion from a 2D domain [of course I can share the Exodus inout + the .i file, if that helps]

Cheers,
Javier

**Comment by garciapintado** [comment]
Argh! Just after writing, I think I’ve found a reason for this segmentation error. This is a multiphysics problem in the sense that the porous flow dictator (within the module PorousFlow) only controls the upper area of an original 2D domain. Then, I do an extrusion to 3D, and the two new ‘front’ and a ‘rear’ domain sidesets are created. On these sides I am using boundary conditions controlled by Porous Flow (PorousFlowOutflowBC), while actually a part of them (towards their bottom) is not a porous media. I am still puzzled about how possibly the extruded 3D model works with <=15 MPI processes, but effectively by removing these BCs the problem disappears. 

I am leaving this still open in case this is an unexpected generic MOOSE behaviour that you might be useful to consider (and out of curiosity in case you have an explanation for this behaviour), but from my practical side now I believe this can be considered solved.

Cheers,
Javier

**Reply by WilkAndy** [reply]
@garciapintado , i also do not understand how this worked for <=15 procs, but failed with >16 procs.  I think we need to understand this.

Is this correct: you had a PorousFlow `BCs` object active on a 2D boundary, but the 3D elements that "owned" this boundary had no PorousFlow `Variables` active on them?

**Reply by garciapintado** [reply]
Correct! These were PorousFlowOutflowBC (for pressure and temperature) on the 2D newly created sides during the 2D to 3D extrusion. The bottom part of the original 2D domain was a block with non porous flow variables [only temperature with a diffusion solver from the thermal conduction module] and so is the bottom block of the 3D domain. Initially, I make the mistake of using the complete new "rear" and "front" for these PorousFlowOutflowBC, and had not realised the issue that part of these were boundary on a block not controlled by PorousFlow. Surprisingly to me now, it has been working well (in two different computers and with various number of MPI processes).

Only when I've moved it to a computer with more processors and I've tried with more MPI processes the model has crashed.

I can't be sure, but it seems to me that with higher number of partitions of the domain among the processes, the chances that  one MPI process has it assigned subdomain with no PorousFlow variable in it may be the reason for the segmentation error to appear now. While if all the MPI processes have at least one area of their 2D boundaries touching a part of the 3D PorousFlow domain, the problem does not show up.

**Reply by GiudGiud** [reply]
My uneducated guess would be that when you moved to more processes you started having a domain with ONLY problematic sides for that BC and still some volumetric dofs for the porous flow variable.
But it s just a guess. You may use Debug/output_process_domains=true to check

**Reply by WilkAndy** [reply]
If what you write is true, @garciapintado and @GiudGiud , i think this will not just happen with PorousFlow, but with other parts of MOOSE.   Do you agree?

**Reply by garciapintado** [reply]
I can't be sure, @WilkAndy. I will try to reproduce the effect tomorrow with a simpler test.

**Comment by garciapintado** [comment]
After the debug advice by @GiudGiud, I am pasting a plot which seems to support the idea that when the subdomain pertaining to one MPI process does not overlap a part of the 3D domain with any Kernel controlled by PorousFlow the segmentation fault appears.
Shown in the plots are the "front" faces of a 3D domain.  To the right-top, the subdomain partition for 24 processes. Right-bottom the domain blocks, where the big bottom red block is  not controlled by Porous Flow [still a "temperature" variable exists in this block as well as in the upper blocks]. This is the result of an extrusion toward this front; such that these blocks project as they are seen toward the rear of the domain. Right-top is the partition for 15 processes. In the 15-proc partition no block is only facing the non-porous-flow domain, which this happens for the 24-proc partition [the lower biggest subdomains to the bottom sides]. For the 15-proc partition, the model runs with these PorousFlowOutflowBC, and the resulting [only one timestep] mass and heat fluxes are shown. It actually can make sense. Mas flux is exactly 0 for the "front" areas facing the non-porous-flow domain, and non 0 in the other top blocks. Heat flux (temperature exists everywhere); assuming it is working fine, should be only picking up the diffusive part of the heat flow. The pattern is weird , but maybe because this is only after 1 tiimestep [the bottom of the domain is hotter].
![bug_BCs_across_multiphysic_domains](https://user-images.githubusercontent.com/44291802/204799517-bd422017-8abf-46d3-8669-b295c9e16ddc.png)
 
@WilkAndy, now I share the feeling that this [the issue of a boundary term trying to access a variable ---porepressure in this case--- only defined in part of its corresponding higher-dimension domain may not not crash and take the 0 value for this non-defined domain; and then crashing if not defined at all] may be a generic MOOSE thing rather than specific to this specific BC.

If so, I would not know what you (MOOSE developers) would prefer as default behaviour {Either leaving it as is; and leave to the user take care of there BCs; or expand the possibility of the BC not crashing and just resulting in 0 for boundaries of domains without the corresponding variable; or just make every stop from the start for mixed situations like this}.

I still am going to try a simpler test to be sure that there is not another thing in my files I am missing...

**Reply by garciapintado** [reply]
Buf! For a synthetic test [two blocks in a quadrangular prism with the lower block not controlled by PorousFlow], trying to reproduce the error, actually MOOSE raises an error about the situation:
*** ERROR ***
'front_water' of type 'PorousFlowOutflowBC' depends on variable(s) 'porepressure'. However, that variable does not appear to be defined on (all of) boundary 'front'.

and a 2D test also raises this error [for the MOOSE version I updated this morning; it does not show for a some-months older version in my iMAC, which just leads to a segmentation error]. But still no error like this is raised for my example above. In both I define porepressure as rectricted to the upper porous blocks. The only difference I can figure out that could make a difference result is that in my real-case scenario, porepressure is initialized from an Exodus file, where the variable is defined everywhere:
```
# Real case - related to the images in the previous post
[Variables]
  [porepressure]
    block = ${porous_blocks}
    family = LAGRANGE
    order = FIRST
    initial_from_file_var = P # in the input Exodus file this is defined everywhere
    initial_from_file_timestep = LATEST
  []
 ...
[]

while in the simpler test I've prepared now, it is initialised with a FunctionIC object from MOOSE
# simpler 2 block case, which effectively MOOSE raises the abovementioned ***ERROR***
[Variables]
  [porepressure]
    block = ${porous_blocks}
    family = LAGRANGE
    order = FIRST
  []
...
[]

[ICs]                                          
  [porepressure_IC]
    block = ${porous_blocks}
    type = FunctionIC
    variable = porepressure
    function = porepressure00    
  []
...
[]
```
Al PorousFlow Kernels are identical in both cases and identically contained to their  corresponding ${porous_blocks}
So, good to know that MOOSE raises this error in the synthetic case, although I continue to be puzzled about my "realistic" case...

**Reply by garciapintado** [reply]
Do you think, @WilkAndy and @GiudGiud that this different initialization of the variable could be the reason for an error of this type:
```
*** ERROR ***
'front_water' of type 'PorousFlowOutflowBC' depends on variable(s) 'porepressure'. However, that variable does not appear to be defined on (all of) boundary 'front'.
```
not being raised by MOOSE when the initialization is from the Exodus file [who has it defined everywhere], despite the use of the variable by is constrained to some part of in indicated boundary within MOOSE?

**Reply by WilkAndy** [reply]
@giudgiud - can you comment on this, please?  It seems that when initialising from the exodus file, MOOSE somehow assigns porepressure everywhere (i wonder does it also calculate material properties everywhere?)

**Reply by GiudGiud** [reply]
why is the Exodus assigned everywhere?
Inside the file you should have a (partial) near the name of the variable. 

Then when you restart it, which method are you using?
In a regular exodus restart you can still have the block restriction
for example
```
[Variables]
  [pore_pressure]
    family = LAGRANGE
    order = first
    initial_from_file_var = pore_pressure
    block = ${some blocks}
  []
[]
```

**Reply by garciapintado** [reply]
Hi @GiudGiud, 
In the input Exodus, the variable is defined everywhere for an entangled reason that probably is not relevant here [related to remeshing and dynamic assignation of the "porous" subdomain in the upper part of the complete domain]. So it is not (partial).

But I do have the block restriction. Please see my pasted block above for the variable declaration in the "real case". Is it possible that the CopyNodalVarsAction for the Exodus-file-based  initialization has something to do with this?

From my user's side, now I can live with this. I should be able define new sidesets contrained to the "porous" areas in the input files and hopefully the problem would be solved for my case. So, I guess the discussion to these different behaviours [apparently depending on initialization ways] is relevant if you consider it so [and I am happy to do some other test if helpful].

**Reply by WilkAndy** [reply]
Yea, i agree with the strategy in your last paragraph, @garciapintado .   When using any PorousFlow BC, most particularly  PorousFlowOutflow, the boundaries should be associated with elements that have porous-flow in them.  Would you like to add a note to this effect in the documentation?

**Reply by garciapintado** [reply]
Hi @WilkAndy, yes a note in the documentation would be nice. If you mean that I personally add the note I don't know how to do it. Should not I be a MOOSE contributor? 
As  as side note, in the output from the two abovementioned models, in Paraview I only see (partial) for elemental variables. In none of these two input [filed-based,  on within MOOSE initialization] no nodal variable shows as partial. Porepressure just shows as 0 for the corresponding 'non-porous' lower block in both domains:
<img width="899" alt="Screenshot 2022-12-01 at 08 02 19" src="https://user-images.githubusercontent.com/44291802/204987091-c7ad9384-5134-4d97-bf49-591b0a1a7fab.png">

**Reply by GiudGiud** [reply]
Hello 

You do not need to be a MOOSE contributor to make a pull request. 
THere's some guidelines here https://mooseframework.inl.gov/framework/contributing.html but overall
- fork the repository
- pull it to your local machine, either through cloning or by adding a new remote (`git remote add <your fork address>`)
- create a new branch from origin/**devel**. Call it my branch (`git checkout -b my_branch`)
- make your change
- `git add` the changed files
- `git commit "my changes, refs <an issue number corresponding to the changes, you may need to create it or you may use an existing porous flow one, as appropriate>"`
- push it online to your fork : `git push origin --set-upstream my_branch`
- go in the PR tab of the main moose repo, and click make a PR

Guillaume

**Reply by garciapintado** [reply]
Ah! OK! Many thanks, @GiudGiud.

BTW, I can now confirm that my original problem is solved by applying the BCs  PorousFlowOutflowBC to only "porous" subdomain boundaries. Now I can use as many processes as physical cores (up to 24 in my machine) without the segmentation fault. 

There is still the unsolved curiosity about why the  mentioned error type as:
*** ERROR ***
'front_water' of type 'PorousFlowOutflowBC' depends on variable(s) 'porepressure'. However, that variable does not appear to be defined on (all of) boundary 'front'.

is not being raised by MOOSE when the initialization of the nodal field 'porepressure' is from the Edoxus input file. But this a very low priority, I guess... 

Javier

**Reply by WilkAndy** [reply]
Maybe not a "very low priority", but "somewhat low priority".   i don't have the time to look into this, unfortunately.

**Comment by garciapintado** [comment, marked as answer]
OK, I've opened an issue (#22877) to add a note to the documentation, and am marking this as answered
